plonky2/src/plonk/conditional_recursive_verifier.rs

```diff
@@ -1,20 +1,65 @@
+use anyhow::{ensure, Result};
 use itertools::Itertools;
 use plonky2_field::extension::Extendable;
+use plonky2_util::ceil_div_usize;
 
 use crate::fri::proof::{
     FriInitialTreeProofTarget, FriProofTarget, FriQueryRoundTarget, FriQueryStepTarget,
 };
 use crate::gadgets::polynomial::PolynomialCoeffsExtTarget;
+use crate::gates::noop::NoopGate;
 use crate::hash::hash_types::{HashOutTarget, MerkleCapTarget, RichField};
 use crate::hash::merkle_proofs::MerkleProofTarget;
 use crate::iop::ext_target::ExtensionTarget;
 use crate::iop::target::{BoolTarget, Target};
+use crate::iop::witness::{PartialWitness, Witness};
 use crate::plonk::circuit_builder::CircuitBuilder;
-use crate::plonk::circuit_data::{CommonCircuitData, VerifierCircuitTarget};
-use crate::plonk::config::{AlgebraicHasher, GenericConfig};
-use crate::plonk::proof::{OpeningSetTarget, ProofTarget, ProofWithPublicInputsTarget};
+use crate::plonk::circuit_data::{CircuitData, CommonCircuitData, VerifierCircuitTarget};
+use crate::plonk::config::{AlgebraicHasher, GenericConfig, Hasher};
+use crate::plonk::proof::{
+    OpeningSetTarget, ProofTarget, ProofWithPublicInputs, ProofWithPublicInputsTarget,
+};
 use crate::with_context;
 
+/// Generate a proof having a given `CommonCircuitData`.
+pub fn dummy_proof<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    common_data: &CommonCircuitData<F, C, D>,
+) -> Result<(ProofWithPublicInputs<F, C, D>, CircuitData<F, C, D>)>
```

Comment:
Couple nits,

- Maybe this should be private; I think we should expose a higher-level method, which looks just like `verify_proof` except with a `BoolTarget` flag, and calls `dummy_proof` internally.
- I would return the `VerifierOnlyCircuitData` part, to avoid redundancy between inputs and outputs (and might be slightly more convenient, if it just gets passed to `conditionally_verify_proof`)

Reply:
> - Maybe this should be private; I think we should expose a higher-level method, which looks just like `verify_proof` except with a `BoolTarget` flag, and calls `dummy_proof` internally.

Hm I'm not sure I understand what that function look like and how the flag would be use. Can you explain a bit more?

> - I would return the `VerifierOnlyCircuitData` part, to avoid redundancy between inputs and outputs (and might be slightly more convenient, if it just gets passed to `conditionally_verify_proof`)

Sounds good!

Reply:
Yeah exactly, I was thinking of a `conditionally_verify_proof` function with only one proof/verifier_data. I think it would be useful for handling variable-size inner proofs. E.g. if need to support three possible sizes, we'd call this new `conditionally_verify_proof` three times, and assert one of the three flags is 1, so the prover would have to use one real VK and two dummy VKs.

Reply:
I agree dummy circuits could be convenient, at least for testing, but I think the typical API for that would look more like
```rust
fn dummy_circuit<...>(config: CircuitConfig, degree_bits: usize) -> CircuitData<...> { ... }
```
I'm not sure if there will be another use case where we already have a `CommonCircuitData` and want to match it? If it's unclear, I'd argue it's best to start private to keep our docs simpler, and since it's easier to change private to public than the reverse.

Reply:
Sounds good, I added a `conditionally_verify_proof_or_dummy` function that takes only one proof and I changed `dummy_proof` to `pub(crate)`.